components/apimgt/org.wso2.carbon.apimgt.gateway/src/main/java/org/wso2/carbon/apimgt/gateway/handlers/streaming/websocket/WebSocketAnalyticsDataProvider.java

```diff
@@ -0,0 +1,283 @@
+
+import java.net.HttpURLConnection;
+import java.util.Map;
+import java.util.UUID;
+
+public class WebSocketAnalyticsDataProvider implements AnalyticsDataProvider {
+    private static final Log log = LogFactory.getLog(WebSocketAnalyticsDataProvider.class);
+
+    ChannelHandlerContext ctx;
+
+    public WebSocketAnalyticsDataProvider(ChannelHandlerContext ctx) {
+        this.ctx = ctx;
+    }
+
+    private AuthenticationContext getAuthenticationContext()  {
+        Object authContext = WebSocketUtils.getPropertyFromChannel(APISecurityUtils.API_AUTH_CONTEXT, ctx);
+        if (authContext != null) {
+            return (AuthenticationContext) authContext;
+        }
+        return null;
+    }
+
+    @Override
+    public EventCategory getEventCategory() {
+        if (isSuccessRequest()) {
+            return EventCategory.SUCCESS;
+        } else if (isFaultRequest()) {
+            return EventCategory.FAULT;
+        } else {
+            return EventCategory.INVALID;
+        }
+    }
+
+    private boolean isSuccessRequest() {
+        return getAuthenticationContext() != null && getErrorCode() == -1;
+    }
+
+    private boolean isFaultRequest() {
+        return getErrorCode() > -1;
+    }
+
+    private int getErrorCode() {
+        Object errorCode = WebSocketUtils.getPropertyFromChannel(SynapseConstants.ERROR_CODE, ctx);
+        if (errorCode != null) {
+            return (int) errorCode;
+        }
+        return -1;
+    }
+
+    @Override
+    public boolean isAnonymous() {
+        AuthenticationContext authContext = getAuthenticationContext();
+        return isAuthenticated() && APIConstants.END_USER_ANONYMOUS.equalsIgnoreCase(authContext.getUsername());
+    }
+
+    @Override
+    public boolean isAuthenticated() {
+        AuthenticationContext authContext = getAuthenticationContext();
+        return authContext != null && authContext.isAuthenticated();
+    }
+
+    @Override
+    public FaultCategory getFaultType() {
+        if (isAuthFaultRequest()) {
+            return FaultCategory.AUTH;
+        } else if (isThrottledFaultRequest()) {
+            return FaultCategory.THROTTLED;
+        } else if (isTargetFaultRequest()) {
+            return FaultCategory.TARGET_CONNECTIVITY;
+        }
+        return FaultCategory.OTHER;
+    }
+
+    private boolean isAuthFaultRequest() {
+        int errorCode = getErrorCode();
+        return errorCode >= Constants.ERROR_CODE_RANGES.AUTH_FAILURE_START
+                && errorCode < Constants.ERROR_CODE_RANGES.AUTH_FAILURE__END;
+    }
+
+    private boolean isThrottledFaultRequest() {
+        int errorCode = getErrorCode();
+        return errorCode >= Constants.ERROR_CODE_RANGES.THROTTLED_FAILURE_START
+                && errorCode < Constants.ERROR_CODE_RANGES.THROTTLED_FAILURE__END;
+    }
+
+    private boolean isTargetFaultRequest() {
+        return false;
+    }
+
+    @Override
+    public API getApi() {
```

Comment:
cant we use the super?

Reply:
We are implementing the `AnalyticsDataProvider` interface here (not the `SynapseAnalyticsDataProvider` class), therefore we can't use super.

Reply:
ok